doc/proposals/openshift-4.3/operator-testing-tool.md

```diff
@@ -0,0 +1,240 @@
+
+Tasks for this story include:
+ * change scorecard by removing earnedPoints and maximumPoints from each test
+ * change scorecard output by removing totalScorePercent, partialPass, earnedPoints, maximumPoints from scorecard output
+ * change scorecard existing tests to remove partialPass, instead only returning a pass or fail
+ * change scorecard CLI to return an exit code based on the pass/fail results from the set of executed tests 
+ * (stretch goal) add support for --fast-fail CLI flag
+
+
+Rough point estimate is 8 points.
+
+
+#### Story 2 - Change Scorecard Test Selection
+Today, the scorecard lets you select which tests are run by the plugins defined in the scorecard configuration (basic, olm). This story would change how users determine which tests are executed by scorecard.
+
+This story would introduce labels for each test and then allow the scorecard CLI to filter which tests it runs based on label selectors.
+
+Tests can fall into 3 groups: required, recommended, or optional. Tests can also be categorized as static or runtime. Labels for these groups and categories would allow a test to be more precisely specified by an end user.
+
+Possible examples of specifying what tests to run are as follows: 
+ * operator-sdk scorecard -l ‘testgroup in (required,recommended,optional)’ 
+ * operator-sdk scorecard -l ‘testtype in (runtime,static)’
+ * operator-sdk scorecard -l updatecrtest=true
+ * operator-sdk scorecard -l someplugintest=true
+
+Tasks for this story include:
+ * Update the test interfaces to support arbitrary test-defined labels
+ * Add the labels to the tests
+ * Update the CLI and scorecard library to support label selectors to filter scorecard runs
+ * A stretch goal would include the ability of the CLI to perform a dry-run of tests based on label selectors to show the user which tests would be executed for a given label selector value.
+
+Rough point estimate for this story is 8 points.
+#### Story 3 - Common Validation
+The scorecard would use a common validation codebase to verify bundle contents. This story supports a single “source of truth” and avoids duplication or variance of validation logic. The scorecard user would be able to specify a bundle on the command line, that bundle would then be validated.  For example:
+ * operator-sdk scorecard --bundle ./my-operator
+
+
+Tasks for this story include:
+ * determine the correct validation library to use
+ * add support for a bundle directory flag to the CLI
+ * change the scorecard code to use that validation library when validating a CR, CSV, or given a bundle directory
+ * format the validation output to present to the CLI user in text or json
+
+Rough point estimate in this story is 5 points.
+
+#### Story 4 - Update scorecard Documentation
+Tasks:
+ * Document changes to CLI
+ * Document new output formats
+ * Document changes to configuration
+ * Document breaking changes and removals
+
+
+
+### Implementation Details/Notes/Constraints
```

Comment:
This feels like the proposal is not implementable. 

1. How are we going to create a labeling system? Does a user provide their own list of labels for a test? do we provide that list? do we provide an initial list? 

Are we planing on writing a DSL of some sort to parse this? `testtype in (runtime,static)` are we worried that we are signing up for more than we bargained for?

Reply:
The idea is to:
1. Allow tests to declare a set of labels that can be introspected by the scorecard at runtime
2. Allow users to filter the set of tests to run based on a label selector string
3. Reuse apimachinery for [parsing](https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L889) and [matching](https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L192).

Reply:
Can we add that info to the Implementation Details of the proposal?

Reply:
I'll update the proposal and include that list in Implementation Details.